pkg/kubelet/stats/cri_stats_provider.go

```diff
@@ -57,10 +65,203 @@ func newCRIStatsProvider(
 func (p *criStatsProvider) ListPodStats() ([]statsapi.PodStats, error) {
-	return nil, fmt.Errorf("not implemented")
+	// Gets node root filesystem information, which will be used to populate
+	// the available and capacity bytes/inodes in container stats.
+	rootFsInfo, err := p.cadvisor.RootFsInfo()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get rootFs info: %v", err)
+	}
+
+	// Creates container map.
+	containerMap := make(map[string]*runtimeapi.Container)
+	containers, err := p.runtimeService.ListContainers(&runtimeapi.ContainerFilter{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list all containers: %v", err)
+	}
+	for _, c := range containers {
+		containerMap[c.Id] = c
+	}
+
+	// Creates pod sandbox map.
+	podSandboxMap := make(map[string]*runtimeapi.PodSandbox)
+	podSandboxes, err := p.runtimeService.ListPodSandbox(&runtimeapi.PodSandboxFilter{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list all pod sandboxes: %v", err)
+	}
+	for _, s := range podSandboxes {
+		podSandboxMap[s.Id] = s
+	}
+
+	// uuidToFsInfo is a map from filesystem UUID to its stats. This will be
+	// used as a cache to avoid querying cAdvisor for the filesystem stats with
+	// the same UUID many times.
+	uuidToFsInfo := make(map[runtimeapi.StorageIdentifier]*cadvisorapiv2.FsInfo)
+
+	// sandboxIDToPodStats is a temporary map from sandbox ID to its pod stats.
+	sandboxIDToPodStats := make(map[string]*statsapi.PodStats)
+
+	resp, err := p.runtimeService.ListContainerStats(&runtimeapi.ListContainerStatsRequest{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list all container stats: %v", err)
+	}
+	for _, stats := range resp.Stats {
+		containerID := stats.Attributes.Id
+		container, found := containerMap[containerID]
+		if !found {
+			glog.Errorf("Unknown id %q in container map: %v", containerID)
+			continue
+		}
+
+		podSandboxID := container.PodSandboxId
+		podSandbox, found := podSandboxMap[podSandboxID]
+		if !found {
+			glog.Errorf("Unknown id %q in pod sandbox map: %v", podSandboxID)
+			continue
+		}
+
+		// Creates the stats of the pod (if not created yet) which the
+		// container belongs to.
+		ps, found := sandboxIDToPodStats[podSandboxID]
+		if !found {
+			ps = p.makePodStats(podSandbox)
+			sandboxIDToPodStats[podSandboxID] = ps
+		}
+		ps.Containers = append(ps.Containers, *p.makeContainerStats(stats, container, &rootFsInfo, uuidToFsInfo))
+	}
+
+	result := make([]statsapi.PodStats, 0, len(sandboxIDToPodStats))
+	for _, s := range sandboxIDToPodStats {
+		result = append(result, *s)
+	}
+	return result, nil
 }
 
+// ImageFsStats returns the stats of the image filesystem.
 func (p *criStatsProvider) ImageFsStats() (*statsapi.FsStats, error) {
-	return nil, fmt.Errorf("not implemented")
+	resp, err := p.imageService.ImageFsInfo(&runtimeapi.ImageFsInfoRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, fs := range resp.ImageFilesystems {
```

Comment:
I forgot to mention this in the PR description. CRI may return the stats of multiple image filesystems but here we only return the first one. Is this an issue?

Reply:
Add a comment here to make it clear.

Reply:
Done.